File: android/feature/auth/src/main/kotlin/com/dualmusic/feature/auth/AuthRepository.kt
package com.dualmusic.feature.auth

import com.dualmusic.core.network.ApiClient
import com.dualmusic.core.network.Endpoint
import com.dualmusic.core.network.TokenStore
import com.dualmusic.domain.auth.AuthEndpoints
import com.dualmusic.domain.auth.AuthSession
import com.dualmusic.domain.auth.ForgotPasswordRequest
import com.dualmusic.domain.auth.LoginRequest
import com.dualmusic.domain.auth.ResetPasswordRequest
import com.dualmusic.domain.auth.MeResponse
import com.dualmusic.domain.auth.RegisterRequest
import com.dualmusic.domain.auth.VerifyOtpRequest
import kotlinx.serialization.json.Json

/**
 * Accès aux opérations d'authentification `/auth/…`.
 *
 * Orchestre [ApiClient] (REST) et [TokenStore] (persistance). Après login/register,
 * persiste la session ; après logout, l'efface.
 *
 * @param baseUrl URL publique SANS `/api/v1` (pour l'URL de démarrage Google).
 */
class AuthRepository(
    private val api: ApiClient,
    private val tokenStore: TokenStore,
    private val baseUrl: String,
) {
    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    /** Connexion email + mot de passe ; persiste la session. */
    suspend fun login(email: String, password: String): AuthSession {
        val body = json.encodeToString(LoginRequest.serializer(), LoginRequest(email, password))
        val session: AuthSession = api.request(Endpoint.post(AuthEndpoints.LOGIN, body).copy(anonymous = true))
        tokenStore.setTokens(session.accessToken, session.refreshToken)
        return session
    }

    /** Inscription ; persiste la session. */
    suspend fun register(request: RegisterRequest): AuthSession {
        val body = json.encodeToString(RegisterRequest.serializer(), request)
        val session: AuthSession = api.request(Endpoint.post(AuthEndpoints.REGISTER, body).copy(anonymous = true))
        tokenStore.setTokens(session.accessToken, session.refreshToken)
        return session
    }

    /** Utilisateur courant (`/auth/me`) — réhydratation au démarrage. */
    suspend fun me(): MeResponse = api.request(Endpoint.get(AuthEndpoints.ME))

    /** Déconnexion : invalide le refresh côté serveur puis efface le stockage local. */
    suspend fun logout() {
        val refresh = tokenStore.refreshToken()
        runCatching {
            api.request<Unit>(Endpoint.post(AuthEndpoints.LOGOUT, """{"refreshToken":${refresh?.let { "\"$it\"" } ?: "null"}}"""))
        }
        tokenStore.clear()
    }

    /** Envoie un OTP au numéro de l'utilisateur connecté. */
    suspend fun sendPhoneOtp() {
        api.request<Unit>(Endpoint.post(AuthEndpoints.OTP_PHONE_SEND))
    }

    /** Vérifie le code OTP (4–8 chiffres). */
    suspend fun verifyPhoneOtp(code: String) {
        val body = json.encodeToString(VerifyOtpRequest.serializer(), VerifyOtpRequest(code))
        api.request<Unit>(Endpoint.post(AuthEndpoints.OTP_PHONE_VERIFY, body))
    }

    /** (Ré)envoie le code de vérification par email au compte connecté. */
    suspend fun sendEmailOtp() {
        api.request<Unit>(Endpoint.post(AuthEndpoints.OTP_EMAIL_SEND))
    }

    /** Vérifie le code email → marque l'email comme vérifié côté serveur. */
    suspend fun verifyEmailOtp(code: String) {
        val body = json.encodeToString(VerifyOtpRequest.serializer(), VerifyOtpRequest(code))
        api.request<Unit>(Endpoint.post(AuthEndpoints.OTP_EMAIL_VERIFY, body))
    }

    /** Demande un email de réinitialisation (renvoie toujours OK, ne révèle rien). */
    suspend fun forgotPassword(email: String) {
        val body = json.encodeToString(ForgotPasswordRequest.serializer(), ForgotPasswordRequest(email.trim()))
        api.request<Unit>(Endpoint.post(AuthEndpoints.PASSWORD_FORGOT, body).copy(anonymous = true))
    }

    /** Réinitialise le mot de passe avec le code reçu par email. */
    suspend fun resetPassword(email: String, code: String, newPassword: String) {
        val body = json.encodeToString(
            ResetPasswordRequest.serializer(),
            ResetPasswordRequest(email.trim(), code.trim(), newPassword),
        )
        api.request<Unit>(Endpoint.post(AuthEndpoints.PASSWORD_RESET, body).copy(anonymous = true))
    }

    /**
     * URL de démarrage OAuth Google (redirect web) — à ouvrir dans un **Custom Tab**.
     * ⚠️ Retour via deep link à gérer ; ou endpoint natif idToken à ajouter côté backend
     * (voir DELIVERY-03).
     */
    fun googleStartUrl(): String = baseUrl.trimEnd('/') + "/api/v1" + AuthEndpoints.OAUTH_GOOGLE_START
}

File: ios/Packages/DualMusicKit/Sources/FeatureAuth/AuthRepository.swift
import Foundation
import CoreNetwork
import DomainModels

/// Accès aux opérations d'authentification `/auth/…`.
///
/// Orchestre ``CoreNetwork/HTTPClient`` (REST) et `TokenStore` (persistance Keychain).
/// Après login/register, persiste la session ; après logout, l'efface.
/// Miroir de `AuthRepository` côté Android.
public struct AuthRepository: Sendable {

    private let http: HTTPClient
    private let tokenStore: TokenStore
    private let baseURL: URL

    /// - Parameters:
    ///   - http: client HTTP applicatif.
    ///   - tokenStore: stockage sécurisé des jetons.
    ///   - baseURL: URL publique **SANS** `/api/v1` (pour l'URL de démarrage Google web).
    public init(http: HTTPClient, tokenStore: TokenStore, baseURL: URL) {
        self.http = http
        self.tokenStore = tokenStore
        self.baseURL = baseURL
    }

    /// Connexion email + mot de passe ; persiste la session.
    /// - Returns: la session (utilisateur, profil, rôles).
    @discardableResult
    public func login(email: String, password: String) async throws -> AuthSession {
        let session: AuthSession = try await http.request(
            .post(AuthEndpoints.login, body: LoginRequest(email: email, password: password), anonymous: true),
            as: AuthSession.self
        )
        await tokenStore.setTokens(access: session.accessToken, refresh: session.refreshToken)
        return session
    }

    /// Inscription ; persiste la session.
    @discardableResult
    public func register(_ request: RegisterRequest) async throws -> AuthSession {
        let session: AuthSession = try await http.request(
            .post(AuthEndpoints.register, body: request, anonymous: true),
            as: AuthSession.self
        )
        await tokenStore.setTokens(access: session.accessToken, refresh: session.refreshToken)
        return session
    }

    /// Connexion Google **native** : envoie l'ID token (obtenu via le SDK Google Sign-In) au
    /// backend qui le vérifie, résout/crée le compte et renvoie une session. La persiste.
    /// - Parameter idToken: ID token signé par Google.
    @discardableResult
    public func loginWithGoogle(idToken: String) async throws -> AuthSession {
        let session: AuthSession = try await http.request(
            .post(AuthEndpoints.oauthGoogleNative, body: GoogleNativeRequest(idToken: idToken), anonymous: true),
            as: AuthSession.self
        )
        await tokenStore.setTokens(access: session.accessToken, refresh: session.refreshToken)
        return session
    }

    /// Utilisateur courant (`GET /auth/me`) — réhydratation au démarrage.
    public func me() async throws -> MeResponse {
        try await http.request(.get(AuthEndpoints.me), as: MeResponse.self)
    }

    /// Complète/édite le profil du caller (`PATCH /users/me`) — étape 3 de l'inscription
    /// (nom, pays, numéro) et édition ultérieure. N'envoie que les champs non nuls.
    public func updateProfile(_ request: UpdateProfileRequest) async throws {
        try await http.send(.patch(UserEndpoints.updateMe, body: request))
    }

    /// Déconnexion : invalide le refresh côté serveur puis efface le stockage local.
    ///
    /// L'échec de l'appel serveur est **volontairement ignoré** : la session locale doit
    /// être effacée dans tous les cas (mode avion, backend indisponible…).
    public func logout() async {
        let refresh = await tokenStore.refreshToken()
        try? await http.send(.post(AuthEndpoints.logout, body: RefreshRequest(refreshToken: refresh)))
        await tokenStore.clear()
    }

    /// Envoie un OTP au numéro de l'utilisateur connecté.
    public func sendPhoneOtp() async throws {
        try await http.send(.post(AuthEndpoints.otpPhoneSend))
    }

    /// Vérifie le code OTP téléphone.
    public func verifyPhoneOtp(code: String) async throws {
        try await http.send(.post(AuthEndpoints.otpPhoneVerify, body: VerifyOtpRequest(code: code)))
    }

    /// (Re)envoie le code de vérification par email au compte connecté.
    public func sendEmailOtp() async throws {
        try await http.send(.post(AuthEndpoints.otpEmailSend))
    }

    /// Vérifie le code email → marque l'email comme vérifié côté serveur.
    public func verifyEmailOtp(code: String) async throws {
        try await http.send(.post(AuthEndpoints.otpEmailVerify, body: VerifyOtpRequest(code: code)))
    }

    /// Demande un email de réinitialisation (le backend répond toujours OK, sans révéler si
    /// le compte existe).
    public func forgotPassword(email: String) async throws {
        try await http.send(
            .post(AuthEndpoints.passwordForgot, body: ForgotPasswordRequest(email: email.trimmed), anonymous: true)
        )
    }

    /// Réinitialise le mot de passe avec le code reçu par email.
    public func resetPassword(email: String, code: String, newPassword: String) async throws {
        try await http.send(
            .post(
                AuthEndpoints.passwordReset,
                body: ResetPasswordRequest(email: email.trimmed, code: code.trimmed, newPassword: newPassword),
                anonymous: true
            )
        )
    }

    /// Enregistre le jeton push de cet appareil auprès du backend (appelé une fois connecté).
    /// - Parameter token: jeton FCM/APNs de l'appareil.
    public func registerDeviceToken(_ token: String) async throws {
        try await http.send(.post(NotificationEndpoints.devices, body: DeviceTokenRequest(token: token)))
    }

    /// URL de démarrage OAuth Google (flux redirect web) — repli si l'échange natif n'est
    /// pas disponible. À ouvrir dans un `ASWebAuthenticationSession`.
    public var googleStartURL: URL {
        baseURL
            .appendingPathComponent("api/v1")
            .appendingPathComponent(AuthEndpoints.oauthGoogleStart.trimmed(charactersIn: "/"))
    }
}
